LJ-423 LJ-127 LJ-267 Upgrade PC to React 19, migrate pages to app router, implement geolocation based actions

lucanovera opened the pull request:
### Description Of Changes

Upgrade the Privacy center to React 19 & NextJS 15. Migrate all pages routes to use app routes and server components. Refactor code to get environment variables. Read geolocation from headers and pass it to api call.

### Code Changes

Upgrade PC to React 19 and NextJS 15
* Updated PC packages
* Added dependencies to the root package.json to fix react 18 and next 14 at the root level
* Migrated SwaggerUI Fides-js docs to the admin-ui project
* Fixed many types

Implement geolocation based action
* Migrates all pages (except fides-js api endpoints) to the new app router using server and client components
* Refactored getPrivacyCenterEnvironment
* Added new env variable to control if the root property (/ path) should try calling the API before reading a local file 


### Steps to Confirm
For full end-to-end of the feature, pull and run the latest Fides Plus. Also, go to the clients folder and run `npm install`

Backwards-compatibility ---
Without making any changes, the privacy center should load from a static config file, keeping existing functionality instact. (If you have added FIDES_PRIVACY_CENTER__USE_API_CONFIG to your .env, remove it)
1.  Open Privacy Center at http://localhost:3001/
2. Check the title is Take control of your data
3. Access and Erasure action are still available

Consent
1. Go to Consent Experiences in admin-ui
2. Make sure the default "US Privacy Center" consent experience that:
- Experience is enabled 
- California is one of the enabled locations 
- at least one privacy notice that is enabled.
- No properties are selected

3. Visit http://localhost:3001/?geolocation=us-ca
4. Click on Manage your consent, fill the email and click continue
5. You should be redirected to the consent preference screen and that you're able to manage consent
6. Visiting a location that isn't enabled  (http://localhost:3001/?geolocation=us-ny) should not allow you to manage consent and show you the "Consent management unavailable" modal instead


Setting up API configs for properties:
1. Go to swagger http://localhost:8080/docs#/Plus/create_property_api_v1_plus_property_post
3. Use the POST Property endpoint to post these 2 new properties:
(If you have created them already, you may skip this step or remove the properties before adding them again)
[purple-property.json](https://github.com/user-attachments/files/18970112/purple-property.json)
[root-property.json](https://github.com/user-attachments/files/18970176/root-property.json)

Geolocated visits to a property with specific path
1. Now visit the PC for that Property: http://localhost:3001/purple
2. Check the title says Take control of your purple data
5. Check the letters are purple
7. Check the only action (besides consent) is "Explicit fallback erasure" 
8. Check that visiting from Argentina (http://localhost:3001/purple?geolocation=ar) the available action is also "Explicit fallback erasure"
9. Check visiting from Utah (http://localhost:3001/purple?geolocation=us-ut) that the available action is US National Access Policy
10. Check visiting from California (http://localhost:3001/purple?geolocation=us-ca) that the available action is California Access Policy
11. Check that visiting from the Netherlands (http://localhost:3001/purple?geolocation=nl) the available action is France and netherlands access policy
12. Check that visiting from Spain (http://localhost:3001/purple?geolocation=es) the available action is Access Policy for EEA

Geolocated visits to a property with a root path
1. Visit the PC: http://localhost:3001/
2. Confirm it is still loading from config file, the title should be "Take control of your data"
3. Now go to clients/privacy-center/.env and add the following `FIDES_PRIVACY_CENTER__USE_API_CONFIG=true`
4. Reload the page
5. Title should now be "This is the root property config"
6. Texts should be in a blue-ish color
7. Check the only action (besides consent) is "Explicit fallback erasure"
8. Check visiting from Utah (http://localhost:3001/?geolocation=us-ut) that the available action is US National Access Policy
9. Check that visiting from Spain (http://localhost:3001/?geolocation=es) the available action is Access Policy for EEA



Reading of Cloudfront Headers
1. To simulate cdn geolocation headers we can install a browser extension like "Modify Header Value" (https://addons.mozilla.org/en-US/firefox/addon/modify-header-value/)
5. After opening the PC, open the extension and set up the `cloudfront-viewer-country` to a value like `ES`
![Screenshot 2025-02-25 at 15-12-47 ](https://github.com/user-attachments/assets/abd42da2-a737-4b53-9859-ffcb3ef311e2)
6. Reload the page, it should now show the "Access policy" for EEA because it's reading the location from the headers
7. Open the developer tools in the Network tab
8. Reload the page
9. Check that there isn't any /location request




### Pre-Merge Checklist

* [x] Issue requirements met
* [ ] All CI pipelines succeeded
* [x] `CHANGELOG.md` updated
  * [ ] Add a https://github.com/ethyca/fides/labels/db-migration label to the entry if your change includes a DB migration
  * [ ] Add a https://github.com/ethyca/fides/labels/high-risk label to the entry if your change includes a high-risk change (i.e. potential for performance impact or unexpected regression) that should be flagged
* Followup issues:
  * [ ] Followup issues created (include link)
  * [ ] No followup issues
* Database migrations:
  * [ ] Ensure that your downrev is up to date with the latest revision on `main`
  * [ ] Ensure that your `downgrade()` migration is correct and works
    * [ ] If a downgrade migration is not possible for this change, please call this out in the PR description!
  * [x] No migrations
* Documentation:
  * [ ] Documentation complete, [PR opened in fidesdocs](https://github.com/ethyca/fidesdocs/pulls)
  * [x] Documentation [issue created in fidesdocs](https://github.com/ethyca/fidesdocs/issues/new/choose)
  * [ ] If there are any new client scopes created as part of the pull request, remember to update public-facing documentation that references our scope registry
  * [ ] No documentation updates required

lucanovera commented on clients/privacy-center/app/server-utils/getPrivacyCenterEnvironment.ts:
```diff
@@ -0,0 +1,86 @@
+"use server";
```
By default, in the app router every component is a server component, but I think being more explicit helps when getting used to think about client vs server components.

lucanovera commented on clients/privacy-center/app/server-utils/getPrivacyCenterEnvironment.ts:
```diff
@@ -0,0 +1,86 @@
+ * @param propertyPath If the property path is provided, the property will be fetched from the API. eg. /myproperty
+ * @param searchParams The search params from the request, required to allow geolocation overrides through query params eg ?geolocation=us-ca
+ */
+const getPrivacyCenterEnvironment = async ({
```
I refactored this a bit to make it as simple as possible since it get's used a lot

lucanovera commented on clients/privacy-center/app/server-utils/getPrivacyCenterEnvironment.ts:
```diff
@@ -0,0 +1,86 @@
+// Cache the environment to avoid re calculating it when getPrivacyCenterEnvironmentCached
+// is called multiple times. The cache only applies to the current request.
+// ref: https://react.dev/reference/react/cache
+const getPrivacyCenterEnvironmentCached = cache(getPrivacyCenterEnvironment);
```
The caching function from React 19 is used to cache the result. This way getPrivacyCenterEnvironmentCached can be called a lot (like in layout, pages, to generate the metadata, etc) but it is still performant.

lucanovera commented on clients/privacy-center/components/Providers.tsx:
```diff
@@ -0,0 +1,54 @@
+"use client";
```
Starting with Providers this becomes a client component, but it gets it's prop directly from the server component PageLayout.tsx.

In the future, we can probably refactor and remove all these stores and just use server and client components.

lucanovera commented on clients/privacy-center/next.config.js:
```diff
@@ -28,15 +19,8 @@ const nextConfig = {
-    }
-    return config;
-  },
+  transpilePackages: ["react-syntax-highlighter", "fidesui"],
```
next-transpile-modules no longer needed after next 13

gilluminate reviewed (commented):
Seeing a few issues in my testing.

* Utah and Spain are both showing the Explicit fallback erasure instead of their respective actions.
* Visiting from a root path does not display "Take control of your purple data"
* I am not showing the "Access policy" for EEA after updating my headers with the extension
* I am also seeing some strange console log warnings when running Privacy Center locally after taking a long time to load
<img width="638" alt="image" src="https://github.com/user-attachments/assets/0aee42cb-4caa-47eb-a3ab-d6665abdf58c" />

lucanovera commented:
> Seeing a few issues in my testing.
> 
> * Utah and Spain are both showing the Explicit fallback erasure instead of their respective actions.
> * Visiting from a root path does not display "Take control of your purple data"
> * I am not showing the "Access policy" for EEA after updating my headers with the extension
> * I am also seeing some strange console log warnings when running Privacy Center locally after taking a long time to load
> 
> <img alt="image" width="638" src="https://private-user-images.githubusercontent.com/103820/416880648-0aee42cb-4caa-47eb-a3ab-d6665abdf58c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDIzMDQ1NzAsIm5iZiI6MTc0MjMwNDI3MCwicGF0aCI6Ii8xMDM4MjAvNDE2ODgwNjQ4LTBhZWU0MmNiLTRjYWEtNDdlYi1hM2FiLWQ2NjY1YWJkZjU4Yy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMzE4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDMxOFQxMzI0MzBaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0yZjBiZjFlMzFjZjJkMjkxZjFlM2VhMGYxZTcwZGYyNzEwZmJjZjY0NDY0ODVmY2FkY2UyZjkwMjE4MDlhM2ZmJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.5J5mGNCOIzKjE3GD2092rouX0qsbe_NlOV5nfwI6RpM">

This is now ready for a re-review. I've changed the /consent page to be exclusively a client-component. This solved many issues including long wait times for getting redirected to the consent page. It can still feel a little slow running it in dev mode, but it's way faster on a production build (npm run build / npm run start).

I've corrected a mistake in my steps to reproduce, using the root property the title should read "This is the root property config".
If you're still seeing the wrong Actions for the location, let me know and we can review it together in a call. I've gone through all the steps and it's working correctly for me. I think the console warning should be gone too, I'm not seeing any.

gilluminate commented:
Overall looking great! Just have a few minor comments and questions

gilluminate commented:
@lucanovera one other note, there seem to be changes to the fides-js docs when I run that build, so be sure to check those and commit them as well 👍

lucanovera commented on clients/privacy-center/components/ClientMetadata.tsx:
```diff
@@ -0,0 +1,16 @@
+}) => (
+  <>
+    <title>{title}</title>
+    <meta name="description" content={description} />
```
You're right. The consent page is the only one that uses ClientMetadata and it can't be accessed directly so I'll remove the description.

lucanovera commented on clients/privacy-center/app/server-environment.ts:
```diff
@@ -118,7 +122,7 @@ const loadConfigFile = async (
       }
       const file = await fsPromises.readFile(path || url, "utf-8");
-      fidesDebugger(`Loaded configuration file: ${urlString}`);
+      console.debug(`Loaded configuration file: ${urlString}`);
```
Apparently it wasn't, I was able to rever it to just using fidesDebugger and it works fine. I must have had another problem at the time I commited the change.

lucanovera replied:
I think it's an issue after the nextjs upgrade, but fidesDebugger will work fine in dev mode but will throw an error during prerendering if you try to do a production build. I will create a ticket to track this.

lucanovera commented on clients/privacy-center/app/server-utils/debugLogServer.ts:
```diff
@@ -0,0 +1,3 @@
+// eslint-disable-next-line no-console
+const debugLogServer = console.debug;
+export default debugLogServer;
```
Good callout! I just removed this in favor of just using fidesDebugger and it works great.

lucanovera replied:
I think after the nextjs upgrade something broke in the way fidesDebugger gets defined and it stopped working during prerendering, throwing this error. 
<img width="749" alt="Captura de pantalla 2025-03-24 a la(s) 1 25 59 p  m" src="https://github.com/user-attachments/assets/0290b881-22db-4ff4-896e-cd1b14eb2801" />

I've created a Tech Debt ticket to track and will add a comment in the code as well. Once the issue is resolved we can use fidesDebugger again.

lucanovera commented:
@gilluminate Thanks for your feedback. I've applied some suggestions, updated the branch and commited the fides-js docs updates. I should be using the fidesDebugger helper you added, but there's an issue with it after the upgrade. I've created the ticket to followup and fix it.

gilluminate reviewed (approved):
This is looking good! Thanks for making those changes. I look forward to see what you find out about fidesDebugger, thanks for creating a follow up ticket for that.
